filetracker/servers/storage.py

```diff
@@ -0,0 +1,243 @@
+        self.base_dir = base_dir
+        self.blobs_dir = os.path.join(base_dir, 'blobs')
+        self.links_dir = os.path.join(base_dir, 'links')
+        self.locks_dir = os.path.join(base_dir, 'locks')
+
+        self.db_path = os.path.join(base_dir, 'metadata.db')
+        self.db = bsddb3.hashopen(self.db_path)
+
+        _makedirs(self.blobs_dir)
+        _makedirs(self.links_dir)
+        _makedirs(self.locks_dir)
+        _makedirs(self.db_path)
+
+    def store(self, name, data, version, size=0, compressed=False, digest=None):
+        """Adds a new file to the storage.
+        
+        If the file with the same name existed before, it's not
+        guaranteed that the link for the old version will exist until
+        the operation completes, but it's guaranteed that the link
+        will never point to an invalid blob.
+
+        Args:
+            name: name of the file being stored.
+                May contain slashes that are treated as path separators.
+            data: binary file-like object with file contents. 
+                Files with unknown length are supported for compatibility with
+                WSGI interface: ``size`` parameter should be passed in these
+                cases.
+            version: new file "version"
+                Link modification time will be set to this timestamp. If
+                the link exists, and its modification time is higher, the
+                file is not overwritten.
+            size: length of ``data`` in bytes
+                If not 0, this takes priority over internal ``data`` size.
+            compressed: whether ``data`` is gzip-compressed
+                If True, the compression is skipped, and file is written as-is.
+                Note that the current server implementation sends
+                'Content-Encoding' header anyway, mandating client to
+                decompress the file.
+            digest: SHA256 digest of the file before compression
+                If specified, the digest will not be computed again, saving
+                resources.
+        """
+        with _exclusive_lock(self._lock_path('links', name)):
+            link_path = self._link_path(name)
+            if os.path.exists(link_path) and _file_version(link_path) > version:
+                return
+
+            # Path to temporary file that may be created in some cases.
+            temp_file_path = None
+
+            if digest is None:
+                # Write data to temp file and calculate hash.
+                temp_file_fd, temp_file_path = tempfile.mkstemp()
+                temp_file = os.fdopen(temp_file_fd, 'wb')
+                _copy_stream(data, temp_file, size)
+                temp_file.close()
+
+                if compressed:
+                    # If data was already compressed, we have to decompress it
+                    # before calculating the digest.
+                    raw_file_fd, raw_file_path = tempfile.mkstemp()
```

Comment:
We leak a temporary file here in case of exception from here.
Also: I think it's not needed to actually create the temporary file, with _file_digest supporting file-like objects.